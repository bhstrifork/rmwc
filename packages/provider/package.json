{
  "name": "@rmwc/provider",
  "version": "0.0.1",
  "description": "RMWC Provider component",
  "main": "./index.js",
  "types": "./index.d.ts",
  "exports": {
    ".": {
      "import": "./index.mjs",
      "require": "./index.js"
    }
  },
  "peerDependencies": {
    "react": ">=16.8.x",
    "react-dom": ">=16.8.x"
  },
  "dependencies": {
    "@rmwc/base": "^14.0.1-alpha.0",
    "@rmwc/types": "^14.0.1-alpha.0"
  }
}
