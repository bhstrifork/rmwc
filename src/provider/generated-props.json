{
  "npm": {},
  "typescript": {
    "useProviderContext": {
      "fileName": "src/provider/index.tsx",
      "flags": {
        "isExternal": false,
        "isOptional": false,
        "isPrivate": false,
        "isProtected": false,
        "isPublic": false,
        "isRest": false,
        "isStatic": false
      },
      "kind": "method",
      "name": "useProviderContext",
      "sourceUrl": "https://github.com/rmwc/rmwc/blob/master/src/provider/index.tsx#L82",
      "signatures": [
        {
          "fileName": "src/provider/index.tsx",
          "kind": "signature",
          "name": "useProviderContext",
          "sourceUrl": "https://github.com/rmwc/rmwc/blob/master/src/provider/index.tsx#L82",
          "parameters": [],
          "returnType": "RMWCProviderProps",
          "type": "() => RMWCProviderProps"
        }
      ]
    },
    "RMWCProvider": {
      "documentation": {
        "contents": [
          "<p>A provider for setting global options in RMWC.</p>\n"
        ],
        "contentsRaw": "A provider for setting global options in RMWC.",
        "metadata": {}
      },
      "fileName": "src/provider/index.tsx",
      "flags": {
        "isExternal": false,
        "isOptional": false,
        "isPrivate": false,
        "isProtected": false,
        "isPublic": false,
        "isRest": false,
        "isStatic": false
      },
      "kind": "method",
      "name": "RMWCProvider",
      "sourceUrl": "https://github.com/rmwc/rmwc/blob/master/src/provider/index.tsx#L85",
      "signatures": [
        {
          "documentation": {
            "contents": [
              "<p>A provider for setting global options in RMWC.</p>\n"
            ],
            "contentsRaw": "A provider for setting global options in RMWC.",
            "metadata": {}
          },
          "fileName": "src/provider/index.tsx",
          "kind": "signature",
          "name": "RMWCProvider",
          "sourceUrl": "https://github.com/rmwc/rmwc/blob/master/src/provider/index.tsx#L85",
          "parameters": [
            {
              "flags": {
                "isExternal": false,
                "isOptional": false,
                "isPrivate": false,
                "isProtected": false,
                "isPublic": false,
                "isRest": false,
                "isStatic": false
              },
              "kind": "parameter",
              "name": "__namedParameters",
              "type": ""
            }
          ],
          "returnType": "Element<>",
          "type": "(__namedParameters: ) => Element<>"
        }
      ]
    },
    "Readme": {
      "fileName": "src/provider/readme.tsx",
      "flags": {
        "isExternal": false,
        "isOptional": false,
        "isPrivate": false,
        "isProtected": false,
        "isPublic": false,
        "isRest": false,
        "isStatic": false
      },
      "kind": "method",
      "name": "Readme",
      "sourceUrl": "https://github.com/rmwc/rmwc/blob/master/src/provider/readme.tsx#L9",
      "signatures": [
        {
          "fileName": "src/provider/readme.tsx",
          "kind": "signature",
          "name": "Readme",
          "sourceUrl": "https://github.com/rmwc/rmwc/blob/master/src/provider/readme.tsx#L9",
          "parameters": [],
          "returnType": "Element<>",
          "type": "() => Element<>"
        }
      ]
    },
    "RMWCProviderProps": {
      "documentation": {
        "contents": [
          "<p>A provider for setting global options in RMWC.</p>\n"
        ],
        "contentsRaw": "A provider for setting global options in RMWC.",
        "metadata": {}
      },
      "fileName": "src/provider/index.tsx",
      "flags": {
        "isExternal": false,
        "isOptional": false,
        "isPrivate": false,
        "isProtected": false,
        "isPublic": false,
        "isRest": false,
        "isStatic": false
      },
      "kind": "interface",
      "name": "RMWCProviderProps",
      "sourceUrl": "https://github.com/rmwc/rmwc/blob/master/src/provider/index.tsx#L47",
      "methods": [],
      "properties": [
        {
          "documentation": {
            "contents": [
              "<p>Children to render</p>\n"
            ],
            "contentsRaw": "Children to render",
            "metadata": {}
          },
          "fileName": "src/provider/index.tsx",
          "flags": {
            "isExternal": false,
            "isOptional": true,
            "isPrivate": false,
            "isProtected": false,
            "isPublic": false,
            "isRest": false,
            "isStatic": false
          },
          "kind": "property",
          "name": "children",
          "sourceUrl": "https://github.com/rmwc/rmwc/blob/master/src/provider/index.tsx#L57",
          "type": "React.ReactNode"
        },
        {
          "documentation": {
            "contents": [
              "<p>Global options for icons</p>\n"
            ],
            "contentsRaw": "Global options for icons",
            "metadata": {}
          },
          "fileName": "src/provider/index.tsx",
          "flags": {
            "isExternal": false,
            "isOptional": true,
            "isPrivate": false,
            "isProtected": false,
            "isPublic": false,
            "isRest": false,
            "isStatic": false
          },
          "kind": "property",
          "name": "icon",
          "sourceUrl": "https://github.com/rmwc/rmwc/blob/master/src/provider/index.tsx#L51",
          "type": "Partial<RMWC.IconOptions>"
        },
        {
          "documentation": {
            "contents": [
              "<p>Enable / Disable interaction ripples globally</p>\n"
            ],
            "contentsRaw": "Enable / Disable interaction ripples globally",
            "metadata": {}
          },
          "fileName": "src/provider/index.tsx",
          "flags": {
            "isExternal": false,
            "isOptional": true,
            "isPrivate": false,
            "isProtected": false,
            "isPublic": false,
            "isRest": false,
            "isStatic": false
          },
          "kind": "property",
          "name": "ripple",
          "sourceUrl": "https://github.com/rmwc/rmwc/blob/master/src/provider/index.tsx#L49",
          "type": "undefined | false | true"
        },
        {
          "documentation": {
            "contents": [
              "<p>Global tooltip options</p>\n"
            ],
            "contentsRaw": "Global tooltip options",
            "metadata": {}
          },
          "fileName": "src/provider/index.tsx",
          "flags": {
            "isExternal": false,
            "isOptional": true,
            "isPrivate": false,
            "isProtected": false,
            "isPublic": false,
            "isRest": false,
            "isStatic": false
          },
          "kind": "property",
          "name": "tooltip",
          "sourceUrl": "https://github.com/rmwc/rmwc/blob/master/src/provider/index.tsx#L53",
          "type": "Partial<TooltipOptions>"
        },
        {
          "documentation": {
            "contents": [
              "<p>Global typography options</p>\n"
            ],
            "contentsRaw": "Global typography options",
            "metadata": {}
          },
          "fileName": "src/provider/index.tsx",
          "flags": {
            "isExternal": false,
            "isOptional": true,
            "isPrivate": false,
            "isProtected": false,
            "isPublic": false,
            "isRest": false,
            "isStatic": false
          },
          "kind": "property",
          "name": "typography",
          "sourceUrl": "https://github.com/rmwc/rmwc/blob/master/src/provider/index.tsx#L55",
          "type": "Partial<TypographyOptions>"
        }
      ]
    }
  }
}
