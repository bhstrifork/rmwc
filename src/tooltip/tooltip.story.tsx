import React from 'react';

import { Tooltip } from './';
import { Avatar } from '../avatar';

export default {
  title: 'Tooltips'
};

export const _Tooltip = () => (
  <Tooltip content="Test Tooltip">
    <a href="#" style={{ margin: '4rem', display: 'inline-block' }}>
      hover
    </a>
  </Tooltip>
);

export const TooltipWithRichContent = () => (
  <Tooltip
    content={
      <div>
        <Avatar size="xsmall" name="James Friedman" /> James Friedman
      </div>
    }
  >
    <a href="#" style={{ margin: '4rem', display: 'inline-block' }}>
      Rich Content
    </a>
  </Tooltip>
);

TooltipWithRichContent.story = {
  name: 'Tooltip with rich content'
};
