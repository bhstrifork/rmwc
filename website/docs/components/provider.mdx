---
sidebar_label: Provider
title: ''
---

import Provider from '@rmwc/provider/README.mdx';

<Provider />
