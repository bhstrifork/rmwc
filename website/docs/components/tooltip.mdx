---
sidebar_label: Tooltip
title: ''
---

import Tooltip from '@rmwc/tooltip/README.mdx';

<Tooltip />
