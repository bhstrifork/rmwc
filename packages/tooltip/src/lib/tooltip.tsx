import React from 'react';
import * as RMWC from '@rmwc/types';
import { classNames, createComponent, Tag } from '@rmwc/base';
import { useProviderContext } from '@rmwc/provider';
import { useToolTipFoundation } from './foundation';
import { AnchorBoundaryType, CssClasses } from '@material/tooltip';

export type TooltipActivationT = 'hover' | 'click' | 'focus';

/** A Tooltip component for displaying informative popover information. */
export interface TooltipProps {
  /** The overlay for the tooltip. */
  overlay: React.ReactNode;
  /** The children that the tooltip belongs to. Must be a single React element. */
  children: React.ReactNode;
  /** Activate the tooltip through one or more interactions. Defaults to `['hover', 'focus']`. */
  activateOn?: TooltipActivationT | TooltipActivationT[];
  /** Custom className to add to the tooltip overlay container. */
  className?: string;
  /** Delay in milliseconds before showing the tooltip when interacting via touch or mouse. */
  enterDelay?: number;
  /** Delay in milliseconds before hiding the tooltip when interacting via touch or mouse. */
  leaveDelay?: number;
  /** Manually control the open state */
  open?: boolean;
  /** Whether or not to show an arrow on the Tooltip. Defaults to `false`. */
  showArrow?: boolean;
  /** Specify whether the anchor element is bounded (element has an identifiable boundary such as a button) or unbounded (element does not have a visually declared boundary such as a text link). */
  anchorBoundaryType?: AnchorBoundaryType;
  /** Specify whether tooltip should be persistent. Persistent tooltip are triggered by clicks. */
  isPersistent?: boolean;
}

export type TooltipHTMLProps = RMWC.HTMLProps<HTMLDivElement>;

/** A Tooltip component for displaying informative popover information. */
export const Tooltip: RMWC.ComponentType<
  TooltipProps,
  TooltipHTMLProps,
  'div'
> = createComponent<TooltipProps, TooltipHTMLProps>(function Tooltip(
  props,
  ref
) {
  const providerContext = useProviderContext();

  const { anchorEl, rootEl, uniqueId } = useToolTipFoundation(props);

  // merge together provider options
  const {
    isPersistent,
    showArrow = false,
    open,
    overlay
  } = {
    ...providerContext.tooltip,
    ...props
  };

  const isRich = typeof overlay !== 'string';

  const className = classNames('mdc-tooltip', {
    'mdc-tooltip--shown': open,
    'mdc-tooltip--rich': isRich,
    'rmwc-tooltip--show-arrow': showArrow
  });

  const child = React.Children.only(props.children);

  if (!React.isValidElement<React.HTMLProps<any>>(child)) {
    return null;
  }

  if (isRich) {
    return (
      <Tag tag="div" className="mdc-tooltip-wrapper--rich">
        {React.cloneElement(child, {
          ...anchorEl.props(child.props),
          ref: anchorEl.reactRef,
          'aria-describedby': uniqueId,
          'data-tooltip-id': uniqueId
        })}
        <Tag
          tag="div"
          className={`${CssClasses.RICH} mdc-tooltip`}
          id={uniqueId}
          role="tooltip"
          aria-hidden="true"
          ref={ref}
          element={rootEl}
          data-mdc-tooltip-persistent={isPersistent}
        >
          <div className="mdc-tooltip__surface mdc-tooltip__surface-animation">
            <div className="mdc-tooltip__content">{overlay}</div>
          </div>
        </Tag>
      </Tag>
    );
  }

  return (
    <>
      <Tag
        tag="div"
        className={className}
        id={uniqueId}
        role="tooltip"
        aria-hidden="true"
        ref={ref}
        element={rootEl}
        data-mdc-tooltip-persistent={open}
      >
        <div className="mdc-tooltip__surface mdc-tooltip__surface-animation">
          {overlay}
        </div>
      </Tag>
      <Tag tag="fragment" element={anchorEl} ref={anchorEl.reactRef}>
        {React.cloneElement(child, {
          ...anchorEl.props(child.props),
          'aria-describedby': uniqueId,
          'data-tooltip-id': uniqueId
        })}
      </Tag>
    </>
  );
});
