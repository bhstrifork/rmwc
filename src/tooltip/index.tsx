import React from 'react';
import { classNames, createComponent, Tag } from '@rmwc/base';
import { useProviderContext } from '@rmwc/provider';
import { useToolTipFoundation } from './foundation';
import { AnchorBoundaryType, CssClasses } from '@material/tooltip';

export type TooltipActivationT = 'hover' | 'click' | 'focus';

/** A Tooltip component for displaying informative popover information. */
export interface TooltipProps {
  /** The overlay content for the tooltip. */
  content: React.ReactNode;
  /** The children that the tooltip belongs to. Must be a single React element. */
  children:
    | React.ReactElement<any, string | React.JSXElementConstructor<any>>
    | undefined;
  /** Activate the tooltip through one or more interactions. Defaults to `['hover', 'focus']`. */
  activateOn?: TooltipActivationT | TooltipActivationT[];
  /** Custom className to add to the tooltip overlay container. */
  className?: string;
  /** Delay in milliseconds before showing the tooltip when interacting via touch or mouse. */
  enterDelay?: number;
  /** Delay in milliseconds before hiding the tooltip when interacting via touch or mouse. */
  leaveDelay?: number;
  /** Manually control the open state */
  open?: boolean;
  /** Whether or not to show an arrow on the Tooltip. Defaults to `false`. */
  showArrow?: boolean;
  /** Specify whether the anchor element is bounded (element has an identifiable boundary such as a button) or unbounded (element does not have a visually declared boundary such as a text link). */
  anchorBoundaryType?: AnchorBoundaryType;
  /** Specify whether tooltip should be persistent. Persistent tooltip are triggered by clicks. */
  isPersistent?: boolean;
}

/** A Tooltip component for displaying informative popover information. */
export const Tooltip = createComponent<TooltipProps>(function Tooltip(
  props,
  ref
) {
  const providerContext = useProviderContext();

  const { anchorEl, rootEl } = useToolTipFoundation(props);

  // merge together provider options
  const {
    isPersistent,
    showArrow = false,
    open
  } = {
    ...providerContext.tooltip,
    ...props
  };

  const isRich = typeof props.content !== 'string';

  const className = classNames('mdc-tooltip', {
    'mdc-tooltip--shown': open,
    'mdc-tooltip--rich': isRich,
    'rmwc-tooltip--show-arrow': showArrow
  });

  const child = React.Children.only(props.children);

  if (!React.isValidElement<React.HTMLProps<any>>(child)) {
    return null;
  }

  if (isRich) {
    return (
      <RichTooltip>
        <Tag element={anchorEl} ref={anchorEl.reactRef}>
          {React.cloneElement(child, {
            ...anchorEl.props(child.props),
            'aria-describedby': 'tooltip-id',
            'data-tooltip-id': 'tooltip-id'
          })}
        </Tag>
        <Tag
          tag="div"
          className={`${CssClasses.RICH} mdc-tooltip`}
          id="tooltip-id"
          role="tooltip"
          aria-hidden="true"
          ref={ref}
          element={rootEl}
          data-mdc-tooltip-persistent={isPersistent && true}
        >
          <div className="mdc-tooltip__surface mdc-tooltip__surface-animation">
            <div className="mdc-tooltip__content">{props.content}</div>
          </div>
        </Tag>
      </RichTooltip>
    );
  }

  return (
    <>
      <Tag
        tag="div"
        className={className}
        id="tooltip-id"
        role="tooltip"
        aria-hidden="true"
        ref={ref}
        element={rootEl}
        data-mdc-tooltip-persistent={open}
      >
        <div className="mdc-tooltip__surface mdc-tooltip__surface-animation">
          {props.content}
        </div>
      </Tag>
      <Tag tag="fragment" element={anchorEl} ref={anchorEl.reactRef}>
        {React.cloneElement(child, {
          ...anchorEl.props(child.props),
          'aria-describedby': 'tooltip-id',
          'data-tooltip-id': 'tooltip-id'
        })}
      </Tag>
    </>
  );
});

const RichTooltip = React.forwardRef(function RichTooltip(
  props: any,
  ref: React.Ref<any>
) {
  return (
    <Tag
      {...props}
      ref={ref}
      element={props.element}
      className="mdc-tooltip-wrapper--rich"
      tag="div"
    >
      {props.children}
    </Tag>
  );
});
