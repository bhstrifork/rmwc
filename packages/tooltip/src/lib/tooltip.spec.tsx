import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import { RMWCProvider } from '../provider';
import { Tooltip } from './tooltip';

describe('Tooltip', () => {
  it('renders', async () => {
    const { asFragment } = render(
      <Tooltip overlay="tooltip" open>
        <span>test</span>
      </Tooltip>
    );
    userEvent.hover(screen.getByText('test'));
    expect(screen.getByText('tooltip')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('activateOn', () => {
    render(
      <Tooltip overlay="tooltip" activateOn="click">
        <span>test</span>
      </Tooltip>
    );
  });

  it('className', () => {
    render(
      <Tooltip overlay="tooltip" className="my-custom-classname">
        <span>test</span>
      </Tooltip>
    );
  });

  it('enterDelay', () => {
    render(
      <Tooltip overlay="tooltip" enterDelay={1000}>
        <span>test</span>
      </Tooltip>
    );
  });

  it('leaveDelay', () => {
    render(
      <Tooltip overlay="tooltip" leaveDelay={1000}>
        <span>test</span>
      </Tooltip>
    );
  });

  it('works with provider', () => {
    render(
      <RMWCProvider tooltip={{}}>
        <Tooltip overlay="tooltip">
          <span>test</span>
        </Tooltip>
      </RMWCProvider>
    );
  });

  it('can display arrow', () => {
    const { asFragment } = render(
      <RMWCProvider tooltip={{}}>
        <Tooltip overlay="tooltip" showArrow>
          <span>test</span>
        </Tooltip>
      </RMWCProvider>
    );

    expect(screen.getByText('tooltip').parentElement).toHaveClass(
      'rmwc-tooltip--show-arrow'
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('can be rich', () => {
    const { asFragment } = render(
      <RMWCProvider tooltip={{}}>
        {
          <Tooltip overlay={<div>tooltip</div>}>
            <span>test</span>
          </Tooltip>
        }
      </RMWCProvider>
    );

    expect(screen.getByText('test').parentElement).toHaveClass(
      'mdc-tooltip-wrapper--rich'
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('can be rich and persistent', () => {
    const { asFragment } = render(
      <RMWCProvider tooltip={{}}>
        {
          <Tooltip overlay={<div>tooltip</div>} isPersistent>
            <span>test</span>
          </Tooltip>
        }
      </RMWCProvider>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
