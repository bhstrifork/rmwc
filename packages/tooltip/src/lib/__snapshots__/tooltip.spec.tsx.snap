// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip can be rich 1`] = `
<DocumentFragment>
  <div
    class="mdc-tooltip-wrapper--rich"
  >
    <span
      aria-describedby="123456"
      class=""
      data-tooltip-id="123456"
    >
      test
    </span>
    <div
      aria-hidden="true"
      class="mdc-tooltip--rich mdc-tooltip"
      id="123456"
      role="tooltip"
    >
      <div
        class="mdc-tooltip__surface mdc-tooltip__surface-animation"
      >
        <div
          class="mdc-tooltip__content"
        >
          <div>
            tooltip
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Tooltip can be rich and persistent 1`] = `
<DocumentFragment>
  <div
    class="mdc-tooltip-wrapper--rich"
  >
    <span
      aria-describedby="123456"
      class=""
      data-tooltip-id="123456"
    >
      test
    </span>
    <div
      aria-hidden="true"
      class="mdc-tooltip--rich mdc-tooltip"
      data-mdc-tooltip-persistent="true"
      id="123456"
      role="tooltip"
    >
      <div
        class="mdc-tooltip__surface mdc-tooltip__surface-animation"
      >
        <div
          class="mdc-tooltip__content"
        >
          <div>
            tooltip
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Tooltip can display arrow 1`] = `
<DocumentFragment>
  <div
    aria-hidden="true"
    class="mdc-tooltip rmwc-tooltip--show-arrow"
    id="123456"
    role="tooltip"
  >
    <div
      class="mdc-tooltip__surface mdc-tooltip__surface-animation"
    >
      tooltip
    </div>
  </div>
  <fragment
    class=""
  >
    <span
      aria-describedby="123456"
      class=""
      data-tooltip-id="123456"
    >
      test
    </span>
  </fragment>
</DocumentFragment>
`;

exports[`Tooltip renders 1`] = `
<DocumentFragment>
  <div
    aria-hidden="true"
    class="mdc-tooltip mdc-tooltip--shown mdc-tooltip--showing"
    data-mdc-tooltip-persistent="true"
    id="123456"
    role="tooltip"
    style="top: 4px; left: 0px;"
  >
    <div
      class="mdc-tooltip__surface mdc-tooltip__surface-animation"
      style="transform-origin: center top;"
    >
      tooltip
    </div>
  </div>
  <fragment
    class=""
  >
    <span
      aria-describedby="123456"
      class=""
      data-tooltip-id="123456"
    >
      test
    </span>
  </fragment>
</DocumentFragment>
`;
